mentorship iosTests/LoginTests.swift

```diff
@@ -0,0 +1,112 @@
+        
+        // Set mock json and data
+        let mockJSON = LoginModel.LoginResponseData(message: "test message")
+        let mockData = try JSONEncoder().encode(mockJSON)
+
+        // Return data in mock request handler
+        MockURLProtocol.requestHandler = { request in
+            return (HTTPURLResponse(), mockData)
+        }
+        
+        // Set expectation. Used to test async code.
+        let expectation = XCTestExpectation(description: "response")
+        // Make login request and test response data.
+        loginService.login(loginData: LoginModel.LoginUploadData(username: "abcd", password: "abcd")) { resp in
+            XCTAssertEqual(resp.message, mockJSON.message)
+            expectation.fulfill()
+        }
+        wait(for: [expectation], timeout: 1)
+    }
+    
+    // MARK: - ViewModel Tests
+
+    func testLoginButtonDisabledState() {
+        let loginVM = LoginViewModel()
+
+        // MARK: - 1. When fields empty. Disabled state should be true.
+        
+        //set disabled state. Currently, data is empty.
+        var loginDisabledState = loginVM.loginDisabled
+        
+        //Test
+        XCTAssertEqual(loginDisabledState, true)
+        
+        // MARK: - 2. When one filed filled and one empty. Disabled state should be true
+        loginVM.loginData = LoginModel.LoginUploadData(username: "username", password: "")
+        
+        loginDisabledState = loginVM.loginDisabled
+
+        //Test
+        XCTAssertEqual(loginDisabledState, true)
+        
+        // MARK: - 3. When all fields filled. Disabled state should be false.
+        loginVM.loginData = LoginModel.LoginUploadData(username: "username", password: "password")
+        
+        loginDisabledState = loginVM.loginDisabled
+        
+        //Test
+        XCTAssertEqual(loginDisabledState, false)
+    }
+    
+    // MARK: - View Tests (Integration Tests)
+    
+    func testLoginActionInView() throws {
+        // Login Service to inject in view for mock network calls
+        let loginService: LoginService = LoginAPI(urlSession: urlSession)
+
+        // View model for login view
+        let loginVM = LoginViewModel()
+        
+        // Login View
+        let loginView = Login(loginService: loginService, loginViewModel: loginVM)
+        
+        // Set mock json and data
+        let mockJSON = LoginModel.LoginResponseData(message: "test message")
+        let mockData = try JSONEncoder().encode(mockJSON)
+        
+        // Return data in mock request handler
+        MockURLProtocol.requestHandler = { request in
+            return (HTTPURLResponse(), mockData)
+        }
+        
+        let expectation = XCTestExpectation(description: "response")
+        loginView.login()
+        expectation.fulfill()
+        wait(for: [expectation], timeout: 2)
+        
+        XCTAssertEqual(loginVM.loginResponseData.message, mockJSON.message)
+    }
```

Comment:
View test tried, failed. I believe this is because expectation are meant to be used with closure based completion handlers.
This also leads me to the question if we need this at all, since we're already fully testing the view model...

Reply:
One solution is to use DispatchQueue.main.asyncAfter(deadline: .now() + 1) instead of XCTestExpectation, and it works. Is this is a good practice?